File: Solutions/46_Permutations/46_Permutations.js
//Fourth fastest
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permute = function(nums) {
    let result = []
    recursion(queue = [], nums)
    function recursion(queue, array){
        
        if(array.length === 0) return result.push(queue)
        
        for(let i = 0; i < array.length; i++){
            let arr = [...array]
            let q = [...queue]
            q.push(arr[i])
            arr.splice(i,1)
            recursion(q, arr)
        }
    }
    return result
};

//Third fastest
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permute = function(nums) {
    let permutations = []
    if (nums.length === 0 || nums.length === 1) {
        permutations.push(nums)
        return permutations
    }
    let permuter = (array, permutation = []) => {
        if (array.length === 0) {
            permutations.push(permutation)
            return
        } 
        for (var i = 0; i < array.length; i++) {
            let copy = array.slice()
            let newArray = copy.splice(i, 1)
            permuter(copy, permutation.concat(newArray))
        }
    }
    permuter(nums)
    return permutations
};

//Second fastest
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permute = function(nums) {
    const res = [];
    const backtrack = (path) => {
      if(path.length === nums.length){
        res.push(path);
        return ;
      }
      nums.forEach(n => {
        if(path.includes(n)){return ;}
        backtrack(path.concat(n));
      });
    };
    backtrack([]);
    return res;
  };

//Fastest
var permuteHelper = function(nums, startIndex, result) {
    if (startIndex === nums.length - 1) {
        result.push([
            ...nums
        ]);
    } else {
        const start = nums[startIndex];
        for (let i = startIndex; i < nums.length; i++) {
            const current = nums[i];
            
            nums[i] = start;
            nums[startIndex] = current;

            permuteHelper(nums, startIndex + 1, result);

            nums[i] = current;
            nums[startIndex] = start;
        }
    }
};

/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permute = function(nums) {
    const result = [];

    permuteHelper(nums, 0, result);

    return result;
};

File: Solutions/46_Permutations/46_Permutations.rb
#Third fastest
# @param {Integer[]} nums
# @return {Integer[][]}
def permute(nums)
    return generate_permute(nums, 0, Array.new)
end

def generate_permute(nums, index, result)
    result << nums.dup if index == nums.length

    for i in index...nums.length
        swap(nums, index, i)
        generate_permute(nums, index+1, result)
        swap(nums, index, i)
    end
    result
end

def swap(nums, i, j)
    t = nums[i]
    nums[i] = nums[j]
    nums[j] = t
end

#Second fastest
# @param {Integer[]} nums
# @return {Integer[][]}
def permute(nums)
    if nums.length == 1
      return [nums]
    end
    
    permutations = []
    nums.each_with_index do |num, idx|
      permute(nums[0, idx] + nums[idx+1..-1]).each do |permutation|
        permutations << [num] + permutation
      end
    end
    
    permutations
  end

#Fastest
def permute(nums)
  results = []
  backtrack(nums, [], results)
  results
end

def backtrack(list, temp, results)
  if temp.length == list.length
    results << temp.dup
  end
  
  list.each do |n|
    next if temp.include? n
    temp << n
    backtrack(list, temp, results)
    temp.pop
  end
end